### ISSUE TITLE
breaking change: define extensibility controller down to AIDL

### ISSUE THREAD

**Issue by atsushieno** [atsushieno/android-audio-plugin-framework#104]
After working on #103, I noticed that we need another bunch of breaking fundamental changes for extensibility:

- We would still need extensibility mechanism that plugins host can "request" calls e.g. "save preset 0 as "%'#$'%$ (binary)", which is not doable just with data pointers. We need non-local function that host can call into plugin.
- We could implement Presets extension as such, plugin service and client interacting together through binder, but that means we need another channel, which will bring in a lot of future API stability breakage. Extensibility should be realized as a generic channel.
- An extension developer (most of the extensions would be implemented by ourselves) should be able to "message" each other (plugin to host, or host to plugin), beyond C API.
- The most likely viable option for "general" purpose messaging is a pair of MIDI controller channels, as detailed at 21edbb9. Then an extension would not even need any extra messaging mechanism.
- However, for operations like sending and receiving state blob that's going to be too massive and involves lots of extraneous work (splitting large binary chunk into MDS, which is what MDS is for but still we don't really need it). Therefore it would be better if an extension can make use of some shared pointer between the plugin and the client host.

Therefore, I would tweak the existing AIDL structure and make some incompatible changes:

- Introduce plugin service extensibility API that every extension developer needs to implement for both host client and plugin service.
- Client hosts can invoke plugin service extension functions.
- Plugins can only reference host extension properties through shared memory. Host cannot get notified if a plugin referenced them.
  - An extension can provide inquiry functions ("do/did you (plugin) reference my (host) property XXX?")
- ExtensionRegistry collects the extension definitions.
- Revamp getState() and setState() AIDL messaging: implement them along with the extensibility framework.

Example Client that makes use of "AAP Presets" extension:

```
	AndroidAudioPluginHost* host;
	auto presetsExt =
		(aap::PresetsPluginServiceExtension*) host->get_extension(AAP_PRESETS_EXTENSION_URI);
	auto preset = presetsExt->get_preset(0);
```

Example Foo Plugin that implements "AAP Presets" extension:

```
	aap_preset_t foo_presets[];
	
	void foo_plugin_factory_initialize() {
		// initialize foo_presets from resources here
	}

	// The actual Presets extension implementation for Foo plugin
	aap_preset_t* foo_presets_get_preset(int32_t index) { return foo_presets[index]; }

	aap_presets_extension_t foo_presets_extension { foo_presets_get_preset };

	void* foo_get_extension(AndroidAudioPlugin *plugin, const char *extensionURI) {
		switch (uri_to_id(extensionURI)) {
		case AAP_PRESETS_EXTENSION_URI_ID:
			return &foo_presets_extension;
		}
	}

	AndroidAudioPlugin foo_plugin{
		nullptr,
		foo_prepare,
		...
		foo_get_extension
	};
	
	AndroidAudioPlugin* foo_plugin_instantiate (...) { return &foo_plugin; };
	
	AndroidAudioPluginFactory foo_factory {
		foo_plugin_instantiate, foo_plugin_release, nullptr
	};
	
	bool foo_factory_initialized{false};
	
	AndroidAudioPluginFactory* GetAndroidAudioPluginFactory () {
		if (!foo_factory_initialized) {
			foo_factory_initialized = true;
			foo_plugin_factory_initialize();
		}
		return &foo_factory;
	}
```

PluginService API (probably C, might be C++):

```
	// C
	void (*aap_service_extension_on_invoked_t) (
		AndroidAudioPluginHost *service,
		AndroidAudioPluginServiceExtension* serviceExtension,
		AndroidAudioPluginExtension* data);

	typedef struct {
		const char *uri;
		aap_service_extension_on_invoked_t on_invoked;
	} aap_service_extension_t;

	// C++
	class AndroidAudioPluginServiceExtension {
		virtual const char* getURI() = 0;
		void clientInvokePluginExtension(AndroidAudioPluginExtension* ext) {
			// transmit data via Binder
		}
		virtual void onServicePluginExtensionInvoked(AndroidAudioPluginExtension* ext) = 0;
	};
```

Presets PluginService extension implementation (probably C, might be C++):

```
	// It is already in aap-core.h
	typedef struct {
		const char *uri;
		int32_t transmit_size;
		void *data;
	} AndroidAudioPluginExtension;

	class MemoryBlock {
		void* buffer{nullptr};
		int32_t size{0};
		
		~MemoryBlock() {
			if (buffer)
				free(buffer);
		}
		
		void copyFrom(int32_t size, void *src) {
			if (buffer == nullptr || this->size < size) {
				if (buffer != nullptr)
					free(buffer);
				this->size = size;
				buffer = calloc(1, size);
			}
			memcpy(result->buffer, *((int32_t*) extensionInstance->data, size);
		}
	}

	// C
	void presets_plugin_service_extension_on_invoked (
		AndroidAudioPluginHost *service,
		AndroidAudioPluginServiceExtension* serviceExtension,
		AndroidAudioPluginExtension* data) {
		switch (opcode) {
		case OPCODE_GET_PRESET:
			int32_t index = *((int32_t*) ext->data);
			auto presetExtension = (aap_preset_extension_t*) service->get_extension(getURI());
			auto preset = (aap_preset_t*) presetExtension->get_preset(index);
			// write size, then data
			*((int32_t*) ext->data) = preset->data_size;
			memcpy(ext->data + sizeof(int32_t), preset->data, preset->data_size);
			return;
		}
		assert(0); // should not reach here			
	}
	
	aap_service_extension_t presets_plugin_service_extension{
		AAP_PRESETS_EXTENSION_URI, presets_plugin_service_extension_on_invoked };
	
	// C++
	class PresetsPluginServiceExtension : public AndroidAudioPluginServiceExtension {
	public:
		PresetsPluginServiceExtension(AndroidAudioPluginHost* host, AndroidAudioPluginExtension *extensionInstance)
			: AndroidAudioPluginServiceExtension(host, extensionInstance) { ... }
	
		inline const char* getURI() override { return AAP_PRESETS_EXTENSION_URI; }
		
		const int32_t OPCODE_GET_PRESET = 0;
		
		// function for client host to implement
		void get_preset(int32_t index, MemoryBlock *result) {
			*((int32_t*) extensionInstance->data) = index;
			clientInvokePluginExtension(data, OPCODE_GET_PRESET);
			int32_t size = *((int32_t*) extensionInstance->data);
			result->copyFrom(size, *((int32_t*) extensionInstance->data + 1);
		}
		
		// function for plugin service to implement, called by AudioPluginService impl.
		void onServicePluginExtensionInvoked(AndroidAudioPluginExtension* ext, int opcode) override {
			switch (opcode) {
			case OPCODE_GET_PRESET:
				int32_t index = *((int32_t*) ext->data);
				auto presetExtension = (aap_preset_extension_t*) host->get_extension(getURI());
				auto preset = (aap_preset_t*) presetExtension->get_preset(index);
				// write size, then data
				*((int32_t*) ext->data) = preset->data_size;
				memcpy(ext->data + sizeof(int32_t), preset->data, preset->data_size);
				return;
			}
			assert(0); // should not reach here			
		}
	};
```

ExtensionRegistry:

```
	class ExtensionRegistry {
		virtual AndroidAudioPluginServiceExtension* create(const char * uri, AndroidAudioPluginHost* host, AndroidAudioPluginExtension *extensionInstance) = 0;
	};
```

Extension service implementation needs to be compiled in the host. For plugins, only the extension header is required (the API has to be "implemented" by the plugin developer anyways).

A host application has to prepare its own `ExtensionRegistry` instance, and manage whatever extensions it wants to support. `StandardExtensionRegistry` should come up with all those standard extensions enabled.

**Comment by atsushieno** [comment]
A fundamental Q occurred today: can we really use two threads for one binder? There is the "binder thread" concept which likely means that a binder is bound to one native thread. We use one audio processing thread, which is not likely realtime yet, but assuming another non-RT thread seems conceptually broken.

Then we would have to revamp the new extension idea of using a dedicated message and run things in the audio thread somehow, and the response must not be a reply message on the binder request as it can be non-RT.

Then we would have to go back to the concept, and resort to MDS etc.

The existing engineering on extensibility API will still take place. But the messaging part will have to change fundamentally.

**Comment by atsushieno** [comment]
as of 3ee772a native implementation part should be mostly done, but since `AudioPluginInstance` does not bind `aap::PluginClient` we cannot simply retrieve AAP extension client proxies and thus cannot test extensions in `androidaudioplugin-ui-compose` yet.

**Comment by atsushieno** [comment]
On updating aap-juce I noticed that we don't provide plugin client (host) API to add host extensions. It is used to provide MidiCIExtension, which provides the native protocol information.

**Comment by atsushieno** [comment]
After some thoughts I realized that host extensions most unlikely makes sense, as AAPXS will be needed. Existing MIDI-CI extension only offers initial channel setup and not really useful and should be removed (it was alive to kind of present to demonstrate extensions, but it does not make sense anymore).

**Comment by atsushieno** [comment]
As of bc751e9 we can instantiate presets extensions from aapinstrumentsample on aap-juce-plugin-host. aap-juce-obxd still causes AudioPluginHost crash (needs more investigation).

**Comment by atsushieno** [comment]
As of aap-juce b736449, aap-juce-obxd and aap-juce-dexed successfully displays list of presets (both show empty-ish data though; they are not likely initialized on Android as those presets depend on local files that the ports are not functional).

**Comment by atsushieno** [comment]
The final API design is somewhat different from the often-updated description here, but the finalized API is being merged to main from `extensions-presets` branch (which should remain for reference) via `extensions-presets-merging` (which would disappear once merged).

**Comment by atsushieno** [comment]
It is now merged as 439e20a.
